src/ReservoirRouter.sol

```diff
@@ -0,0 +1,291 @@
+        address to
+    ) external payable returns (uint256[] memory amounts) {
+        amounts = ReservoirLibrary.getAmountsOut(address(factory), amountIn, path, curveIds);
+        // but the actual swap results might be diff from this. Should we move the require into _swap to check for the minOut?
+        require(amounts[amounts.length - 1] >= amountOutMin, "RL: INSUFFICIENT_OUTPUT_AMOUNT");
+
+        _pay(path[0], msg.sender, ReservoirLibrary.pairFor(address(factory), path[0], path[1], curveIds[0]), amounts[0]);
+        _swap(amounts, true, path, curveIds, to);
+    }
+
+    function swapVariableForExact(
+        uint256 amountOut,
+        uint256 amountInMax,
+        address[] calldata path,
+        uint256[] calldata curveIds,
+        address to
+    ) external payable returns (uint256[] memory amounts) {
+        amounts = ReservoirLibrary.getAmountsIn(address(factory), amountOut, path, curveIds);
+        require(amounts[0] <= amountInMax, "RL: EXCESSIVE_INPUT_AMOUNT");
+
+        _pay(path[0], msg.sender, ReservoirLibrary.pairFor(address(factory), path[0], path[1], curveIds[0]), amounts[0]);
+        _swap(amounts, false, path, curveIds, to);
+    }
+
+    function getAmountOut(
+        uint256 amountIn,
+        uint256 reserveIn,
+        uint256 reserveOut,
+        uint256 curveId,
+        uint256 swapFee,
+        ExtraData calldata extraData
+    ) external pure returns (uint256 amountOut) {
+        if (curveId == 0) {
+            return ReservoirLibrary.getAmountOutConstantProduct(amountIn, reserveIn, reserveOut, swapFee);
+        }
+        else if (curveId == 1) {
+            return ReservoirLibrary.getAmountOutStable(amountIn, reserveIn, reserveOut, swapFee, extraData);
+        }
+    }
+
+    function getAmountIn(
+        uint256 amountOut,
+        uint256 reserveIn,
+        uint256 reserveOut,
+        uint256 curveId,
+        uint256 swapFee,
+        ExtraData calldata extraData
+    ) external pure returns (uint256 amountIn) {
+        if (curveId == 0) {
+            return ReservoirLibrary.getAmountInConstantProduct(amountOut, reserveIn, reserveOut, swapFee);
+        }
+        else if (curveId == 1) {
+            return ReservoirLibrary.getAmountInStable(amountOut, reserveIn, reserveOut, swapFee, extraData);
+        }
+    }
+
+    // perf: to use calldata or memory for path and curveIds?
+    function getAmountsOut(
+        uint256 amountIn,
+        address[] calldata path,
+        uint256[] calldata curveIds
+    ) external view returns(uint256[] memory amountsOut) {
+        return ReservoirLibrary.getAmountsOut(address(factory), amountIn, path, curveIds);
+    }
+
+    function getAmountsIn(
+        uint256 amountOut,
+        address[] calldata path,
+        uint256[] calldata curveIds
+    ) external view returns(uint256[] memory amountsIn) {
+        return ReservoirLibrary.getAmountsIn(address(factory), amountOut, path, curveIds);
+    }
+
+    function quoteAddLiquidity(
```

Comment:
Thoughts on a separate contract for quoting to improve gas efficiency of the router? We can of course revisit this in optimization mode unless we needed to make the decision now to avoid later overhead

Reply:
I had this thought too. To move all the all the view functions into a `Quoter.sol` contract. UniV3 does this. That contract would have a reference to the factory and WETH as well.

Reply:
Do it in a later PR?

Reply:
okay noted.